analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs

```diff
@@ -27,122 +28,110 @@
-                            {
-                                var variableSymbol = GetAccessedVariable(assignmentExpression, c.SemanticModel);
-                                if (variableSymbol != null)
-                                {
-                                    symbolsWhereLocaleIsSet.Add(variableSymbol);
-                                }
-                            }
-                        }, SyntaxKind.SimpleAssignmentExpression);
-
-                    compilationStartContext.RegisterCompilationEndAction(
-                        c =>
-                        {
-                            var invalidDataTypeCreation = symbolsWhereTypeIsCreated
-                                .Where(x => !symbolsWhereLocaleIsSet.Contains(x.Symbol));
-
-                            foreach (var invalidCreation in invalidDataTypeCreation)
-                            {
-                                var typeName = invalidCreation.Symbol.GetSymbolType()?.Name;
-                                if (typeName == null)
-                                {
-                                    continue;
-                                }
-
-                                c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, invalidCreation.Syntax.GetLocation(),
-                                    typeName));
-                            }
-                        });
+                    compilationStartContext.RegisterSyntaxNodeActionInNonGenerated(ProcessObjectCreations(symbolsWhereTypeIsCreated),
+                                                                                   SyntaxKind.ObjectCreationExpression,
+                                                                                   SyntaxKindEx.ImplicitObjectCreationExpression);
+                    compilationStartContext.RegisterSyntaxNodeActionInNonGenerated(ProcessSimpleAssignments(symbolsWhereLocaleIsSet), SyntaxKind.SimpleAssignmentExpression);
+                    compilationStartContext.RegisterCompilationEndAction(ProcessCollectedSymbols(symbolsWhereTypeIsCreated, symbolsWhereLocaleIsSet));
                 });
-        }
 
-        private static SyntaxNode GetAssignmentTargetVariable(ObjectCreationExpressionSyntax objectCreation)
-        {
-            var parent = objectCreation.GetFirstNonParenthesizedParent();
+        private static Action<SyntaxNodeAnalysisContext> ProcessObjectCreations(ConcurrentBag<SyntaxNodeWithSymbol<SyntaxNode, ISymbol>> symbolsWhereTypeIsCreated) =>
+            c =>
+            {
+                if (!(GetSymbolFromConstructorInvocation(c.Node, c.SemanticModel) is ITypeSymbol objectType) || !objectType.IsAny(CheckedTypes))
+                {
+                    return;
+                }
 
-            if (parent is AssignmentExpressionSyntax assignment)
+                var variableSyntax = GetAssignmentTargetVariable(c.Node);
+                if (variableSyntax == null)
+                {
+                    return;
+                }
```

Comment:
While we're touching it, let's refactor this logic to a simple positive and straight forward:
```
                if (GetSymbolFromConstructorInvocation(c.Node, c.SemanticModel) is ITypeSymbol objectType
                    && objectType.IsAny(CheckedTypes)
                    && GetAssignmentTargetVariable(c.Node) is { } variableSyntax)
                {
                    var variableSymbol = variableSyntax is IdentifierNameSyntax
                        ? c.SemanticModel.GetSymbolInfo(variableSyntax).Symbol
                        : c.SemanticModel.GetDeclaredSymbol(variableSyntax);
                    if (variableSymbol != null)
                    {
                        symbolsWhereTypeIsCreated.Add(variableSymbol.ToSymbolWithSyntax(c.Node));
                    }
                }
```